File: app/src/main/java/ashunevich/uniconverter20/Activity_converter.java
package ashunevich.uniconverter20;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.app.Fragment;
import android.text.Editable;


import android.text.TextUtils;
import android.text.TextWatcher;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;

import android.widget.Toast;

import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;

import java.util.HashMap;
import java.util.Locale;

import ashunevich.uniconverter20.databinding.ConverterActivityBinding;

import static ashunevich.uniconverter20.Activity_converter_Utils.blockInput;

public class Activity_converter extends Fragment {

    private ConverterActivityBinding binding;
    private String  sDefSystemLanguage;
    protected HashMap<String, String> hm;

    EventBus bus;
    public Activity_converter() {
        // Required empty public constructor
    }

    @Override
    public void onStart(){
        if (!EventBus.getDefault().isRegistered(this)) { EventBus.getDefault().register(this); }
        super.onStart();
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    @Override
    @Subscribe
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        binding = ConverterActivityBinding.inflate(inflater, container, false);
        View view = binding.getRoot();
        sDefSystemLanguage = Locale.getDefault().getDisplayLanguage();
        bus = EventBus.getDefault();
        blockInput(binding.resultView,binding.valueEdit);
        setAdapter(getResources().getStringArray(R.array.weight));
        addTextWatcher();
        setSpinnersListeners();
        setUnitMeasurement();
        return view;
    }

    private void setAdapter( String [] array ){
        ArrayAdapter<String> adapter = new ArrayAdapter<>(getContext(),
                R.layout.custom_spinner_item,array);
        binding.spinnerValue.setAdapter(adapter);
        binding.spinnerResult.setAdapter(adapter);
    }

    @Override
    public void onStop() {
        EventBus.getDefault().unregister(this);
        super.onStop();
    }

    @Override
    public void onDestroyView() {
        binding = null;
        super.onDestroyView();
    }

    @Subscribe
    public void getText (BusPost_Tab_Position event) {
            setSpinnerAdapterOnBusEvent(event.getPos());
            Activity_converter_Utils.clearView(binding.valueEdit,binding.resultView);
    }

    @Subscribe()
    public void getCalculatorNumber(BusPost_Number event) {
        if (event.getNumber().contains("check") | event.getNumber().contains("correction") |
               event.getNumber().contains("clear")){
           switch (event.getNumber()){
               case "check": Activity_converter_Utils.appendMinusPlus(binding.valueEdit);break;
               case "correction": Activity_converter_Utils.correctValue(binding.valueEdit,binding.resultView);break;
               case "clear": Activity_converter_Utils.clearView(binding.valueEdit,binding.resultView);break;
           }
       }
         else{
            binding.valueEdit.append(event.getNumber());
            }
        }

    //Filling spinners with values
    private void setSpinnerAdapterOnBusEvent(int activeValue) {
            switch (activeValue) {
                case 0: setAdapter(getResources().getStringArray(R.array.weight));  break;
                case 1: setAdapter(getResources().getStringArray(R.array.length)); break;
                case 2: setAdapter(getResources().getStringArray(R.array.volume)); break;
                case 3: setAdapter(getResources().getStringArray(R.array.area)); break;
                case 4: setAdapter(getResources().getStringArray(R.array.force)); break;
                case 5: setAdapter(getResources().getStringArray(R.array.temperature_array)); break;
                case 6: setAdapter(getResources().getStringArray(R.array.time_array)); break;
                case 7: setAdapter(getResources().getStringArray(R.array.speed)); break;
                case 8: setAdapter(getResources().getStringArray(R.array.circlesAndSpheres)); break;
        } }

    //if user changes unit - it will change mesaurments and will automatically recalculate result
    private void setSpinnersListeners(){
        binding.spinnerValue.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> parentView, View selectedItemView, int position, long id) {
                convertAndShowValues(sDefSystemLanguage);
                setUnitMeasurement();
            }
            @Override
            public void onNothingSelected(AdapterView<?> parentView) {

            }

        });

        binding.spinnerResult.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> parentView, View selectedItemView, int position, long id) {
                setUnitMeasurement();
                convertAndShowValues(sDefSystemLanguage);
            }
            @Override
            public void onNothingSelected(AdapterView<?> parentView) {
            }});

    }

    //Auto conversion when user add number to value for convert
    private void addTextWatcher() {
        binding.valueEdit.addTextChangedListener(new TextWatcher() {
                @Override
                public void afterTextChanged(Editable s) {
                    if(binding.valueEdit.getText().toString().trim().length() > 15){
                        binding.valueEdit.setText(s.toString().substring(0, 15));
                        binding.valueEdit.setSelection(s.length()-1);
                        showToast();
                    }
                        convertAndShowValues(sDefSystemLanguage);
                }

                @Override
                public void beforeTextChanged(CharSequence s, int start,
                                              int count, int after) {
                }
                @Override
                public void onTextChanged(CharSequence s, int start,
                                          int before, int count) {

                }
            });
        }

    protected void showToast(){
            Toast.makeText(getActivity(), getResources().getString(R.string.maxNumberReached), Toast.LENGTH_SHORT).show();
        }

    //set units of mesaurments for value
    private void setUnitMeasurement(){
        Activity_converter_Logic.setUnitsView(binding.spinnerValue.getSelectedItem().toString(),binding.valueUnit);
        Activity_converter_Logic.setUnitsView(binding.spinnerResult.getSelectedItem().toString(),binding.resultUnit );
    }

    private void convertAndShowValues(String activeLocale){

        if(TextUtils.isEmpty(binding.valueEdit.getText().toString()) | binding.valueEdit.getText().toString().equals("-")){
            binding.resultView.setText("");
        }
        else {
               if (activeLocale.equals("українська")){
                   Activity_converter_Logic.ConvertValues_Ukr(binding.spinnerValue.getSelectedItem().toString(),
                           binding.spinnerResult.getSelectedItem().toString(),
                           Double.parseDouble(binding.valueEdit.getText().toString()),
                           binding.resultView);
               }
               else{
                   Activity_converter_Logic.ConvertValues(binding.spinnerValue.getSelectedItem().toString(),
                           binding.spinnerResult.getSelectedItem().toString(),
                           Double.parseDouble(binding.valueEdit.getText().toString()),
                           binding.resultView);
               }
           }

        }

    @Override
    public void onDetach() {
        if (EventBus.getDefault().isRegistered(this)) { EventBus.getDefault().unregister(this); }
        super.onDetach();
    }
}

File: app/src/main/java/ashunevich/uniconverterKT/ActivityConverter.kt
package ashunevich.uniconverterKT

import android.os.Bundle
import android.text.Editable
import android.text.TextUtils
import android.text.TextWatcher
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.AdapterView
import android.widget.ArrayAdapter
import android.widget.Spinner
import android.widget.Toast
import androidx.fragment.app.Fragment
import ashunevich.uniconverterKT.Utils.appendMinusPlus
import ashunevich.uniconverterKT.Utils.clearView
import ashunevich.uniconverterKT.Utils.correctValue
import ashunevich.uniconverterKT.databinding.ConverterActivityBinding
import com.ashunevich.conversionlibrary.UnitConverter

open class ActivityConverter : Fragment() {
    private var binding: ConverterActivityBinding? = null

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        // Inflate the layout for this fragment
        binding = ConverterActivityBinding.inflate(inflater, container, false)
        setSpinnerOnTabPositionChange(DEFAULT_POS)
        setUnitMeasurement()
        Utils.blockInput(binding!!.resultView, binding!!.valueEdit)
        return binding!!.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        initTabPositionViewModel()
        initKeyBoardViewModel()
        setSpinnerListener(binding!!.spinnerValue)
        setSpinnerListener(binding!!.spinnerResult)
        addTextWatcher()
        super.onViewCreated(view, savedInstanceState)
    }

    private fun initTabPositionViewModel() {
        val tabPositionViewModel = Utils.generateViewModel(requireActivity())
        tabPositionViewModel.getSelected().observe(
            viewLifecycleOwner
        ) { tabPos: Int -> setSpinnerOnTabPositionChange(tabPos) }
    }

    private fun initKeyBoardViewModel() {
        val keyboardViewModel = Utils.generateViewModel(requireActivity())
        keyboardViewModel.postedNumber.observe(
            viewLifecycleOwner
        ) { event: String? ->
            if (event != null) {
                setViewModelTextReceiver(event)
            }
        }
    }

    private fun setAdapter(array: Array<String>) {
        binding!!.spinnerValue.adapter = ArrayAdapter(
            requireContext(),
            R.layout.custom_spinner_item, array
        )
        binding!!.spinnerResult.adapter = ArrayAdapter(
            requireContext(),
            R.layout.custom_spinner_item, array
        )
    }

    override fun onDestroyView() {
        binding = null
        super.onDestroyView()
    }

    private fun setViewModelTextReceiver(event: String) {
        if (event == Utils.SYMBOL_CORRECT || event == Utils.SYMBOL_CLEAR || event == Utils.SYMBOL_CHECK) {
            when (event) {
                Utils.SYMBOL_CORRECT -> correctValue(
                    binding!!.valueEdit, binding!!.resultView
                )
                Utils.SYMBOL_CLEAR -> clearView(
                    binding!!.valueEdit, binding!!.resultView
                )
                Utils.SYMBOL_CHECK -> appendMinusPlus(
                    binding!!.valueEdit
                )
            }
        } else {
            binding!!.valueEdit.append(event)
        }
    }

    // Filling spinners with values
    private fun setSpinnerOnTabPositionChange(tabPos: Int) {
        when (tabPos) {
            0 -> setAdapter(resources.getStringArray(R.array.weight))
            1 -> setAdapter(resources.getStringArray(R.array.length))
            2 -> setAdapter(resources.getStringArray(R.array.volume))
            3 -> setAdapter(resources.getStringArray(R.array.area))
            4 -> setAdapter(resources.getStringArray(R.array.force))
            5 -> setAdapter(resources.getStringArray(R.array.temperature_array))
            6 -> setAdapter(resources.getStringArray(R.array.time_array))
            7 -> setAdapter(resources.getStringArray(R.array.speed))
        }
    }

    // if user changes unit - it will change measurements and will automatically recalculate result
    private fun setSpinnerListener(spinner: Spinner) {
        spinner.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
            override fun onItemSelected(
                parentView: AdapterView<*>?,
                selectedItemView: View,
                position: Int,
                id: Long
            ) {
                convertAndShowValues(Utils.returnLocale())
                setUnitMeasurement()
            }

            override fun onNothingSelected(parentView: AdapterView<*>?) {}
        }
    }

    // Auto conversion when user add number to value for convert
    private fun addTextWatcher() {
        binding!!.valueEdit.addTextChangedListener(object : TextWatcher {
            override fun afterTextChanged(s: Editable) {
                if (binding!!.valueEdit.text.toString().trim { it <= ' ' }.length > 15) {
                    binding!!.valueEdit.setText(s.toString().substring(0, 15))
                    binding!!.valueEdit.setSelection(s.length - 1)
                    showToast()
                }
                convertAndShowValues(Utils.returnLocale())
            }

            override fun beforeTextChanged(
                s: CharSequence,
                start: Int,
                count: Int,
                after: Int
            ) {
            }

            override fun onTextChanged(
                s: CharSequence,
                start: Int,
                before: Int,
                count: Int
            ) {
            }
        })
    }

    protected fun showToast() {
        Toast.makeText(activity, resources.getString(R.string.maxNumberReached), Toast.LENGTH_SHORT)
            .show()
    }

    // set units of measurements for value
    private fun setUnitMeasurement() {
        Utils.measurementUnitsHandler(
            Utils.getSpinnerValueString(
                binding!!.spinnerValue
            ), binding!!.valueUnit
        )
        Utils.measurementUnitsHandler(
            Utils.getSpinnerValueString(
                binding!!.spinnerResult
            ), binding!!.resultUnit
        )
    }

    private fun convertAndShowValues(activeLocale: String) {
        if (TextUtils.isEmpty(binding!!.valueEdit.text.toString()) or (binding!!.valueEdit.text.toString() == "-")) {
            binding!!.resultView.setText("")
        } else {
            if (activeLocale == "українська") {
                binding!!.resultView.setText(
                    UnitConverter.ConvertValues_Ukr(
                        Utils.getSpinnerValueString(
                            binding!!.spinnerValue
                        ),
                        Utils.getSpinnerValueString(binding!!.spinnerResult),
                        stringToDouble()
                    )
                )
            } else {
                binding!!.resultView.setText(
                    UnitConverter.ConvertValues(
                        Utils.getSpinnerValueString(
                            binding!!.spinnerValue
                        ),
                        Utils.getSpinnerValueString(binding!!.spinnerResult),
                        stringToDouble()
                    )
                )
            }
        }
    }

    private fun stringToDouble(): Double {
        return binding!!.valueEdit.text.toString().toDouble()
    }

    companion object {
        private const val DEFAULT_POS = 0
    }
}